File: fs/py_before_checksum/140__word-break-ii/word-break-ii_benchmark_v2.py
# Time:  O(n * l^2 + n * r), l is the max length of the words,
#                            r is the number of the results.
# Space: O(n^2)

class Solution(object):
    def wordBreak(self, s, wordDict):
        """
        :type s: str
        :type wordDict: Set[str]
        :rtype: List[str]
        """
        n = len(s)

        max_len = 0
        for string in wordDict:
            max_len = max(max_len, len(string))

        can_break = [False for _ in xrange(n + 1)]
        valid = [[False] * n for _ in xrange(n)]
        can_break[0] = True
        for i in xrange(1, n + 1):
            for l in xrange(1, min(i, max_len) + 1):
                if can_break[i-l] and s[i-l:i] in wordDict:
                    valid[i-l][i-1] = True
                    can_break[i] = True

        result = []
        if can_break[-1]:
            self.wordBreakHelper(s, valid, 0, [], result)
        return result

    def wordBreakHelper(self, s, valid, start, path, result):
        if start == len(s):
            result.append(" ".join(path))
            return
        for i in xrange(start, len(s)):
            if valid[start][i]:
                path += [s[start:i+1]]
                self.wordBreakHelper(s, valid, i + 1, path, result)
                path.pop()


if __name__ == "__main__":
    tests = [
        {'s': 'catsanddog', 'dict': ['cat','cats','and','sand','dog']},
        {'s': 'pineapplepenapple', 'dict': ['apple','pen','applepen','pine','pineapple']},
        {'s': 'aaaaaaa', 'dict': ['a','aa','aaa']},
        {'s': 'leetcode', 'dict': ['leet','code']},
        {'s': 'catsandog', 'dict': ['cats','dog','sand','and','cat']},
        {'s': 'abcd', 'dict': ['a','abc','b','cd']},
        {'s': 'aaaaab', 'dict': ['a','aa','aaa','aaaa','b']},
        {'s': 'bb', 'dict': ['a','b','bbb','bbbb']},
        {'s': 'foobarbaz', 'dict': ['foo','bar','baz','foobar','barbaz','foob','arb','az']},
        {'s': 'impossible', 'dict': ['im','possible','impo','ss','ible']}
    ]

    sol = Solution()
    sink = 0

    iterations = 1
    for _ in xrange(iterations):
        checksum = 0
        for t in tests:
            res = sol.wordBreak(t['s'], set(t['dict']))
            checksum += int(len(res))
        sink += checksum
    print(sink)

File: fs/cpp_before_checksum/140__word-break-ii/word-break-ii_benchmark_v2.cpp
#include <iostream>
#include <string>
#include <vector>
#include <unordered_set>
#include <algorithm>

class Solution {
public:
    std::vector<std::string> wordBreak(const std::string& s, std::vector<std::string>& wordDict) {
        int n = static_cast<int>(s.size());
        std::unordered_set<std::string> dict(wordDict.begin(), wordDict.end());

        int max_len = 0;
        for (const auto& w : wordDict) {
            max_len = std::max(max_len, static_cast<int>(w.size()));
        }

        std::vector<bool> can_break(n + 1, false);
        std::vector<std::vector<bool>> valid(n, std::vector<bool>(n, false));
        can_break[0] = true;

        for (int i = 1; i <= n; ++i) {
            for (int l = 1; l <= std::min(i, max_len); ++l) {
                if (can_break[i - l]) {
                    if (dict.find(s.substr(i - l, l)) != dict.end()) {
                        valid[i - l][i - 1] = true;
                        can_break[i] = true;
                    }
                }
            }
        }

        std::vector<std::string> result;
        if (can_break[n]) {
            std::vector<std::string> path;
            wordBreakHelper(s, valid, 0, path, result);
        }
        return result;
    }

private:
    void wordBreakHelper(const std::string& s,
                         const std::vector<std::vector<bool>>& valid,
                         int start,
                         std::vector<std::string>& path,
                         std::vector<std::string>& result) {
        if (start == static_cast<int>(s.size())) {
            result.push_back(join(path));
            return;
        }
        for (int i = start; i < static_cast<int>(s.size()); ++i) {
            if (valid[start][i]) {
                path.push_back(s.substr(start, i - start + 1));
                wordBreakHelper(s, valid, i + 1, path, result);
                path.pop_back();
            }
        }
    }

    std::string join(const std::vector<std::string>& parts) {
        std::string res;
        for (size_t i = 0; i < parts.size(); ++i) {
            if (i) res.push_back(' ');
            res += parts[i];
        }
        return res;
    }
};

struct Test {
    std::string s;
    std::vector<std::string> dict;
};

int main() {
    std::vector<Test> tests = {
        {"catsanddog", {"cat","cats","and","sand","dog"}},
        {"pineapplepenapple", {"apple","pen","applepen","pine","pineapple"}},
        {"aaaaaaa", {"a","aa","aaa"}},
        {"leetcode", {"leet","code"}},
        {"catsandog", {"cats","dog","sand","and","cat"}},
        {"abcd", {"a","abc","b","cd"}},
        {"aaaaab", {"a","aa","aaa","aaaa","b"}},
        {"bb", {"a","b","bbb","bbbb"}},
        {"foobarbaz", {"foo","bar","baz","foobar","barbaz","foob","arb","az"}},
        {"impossible", {"im","possible","impo","ss","ible"}}
    };

    Solution sol;
    volatile int sink = 0;

    const int iterations = 1;
    for (int iter = 0; iter < iterations; ++iter) {
        int checksum = 0;
        for (auto& t : tests) {
            auto res = sol.wordBreak(t.s, t.dict);
            checksum += static_cast<int>(res.size());
        }
        sink += checksum;
        

    } std::cout << sink << std::endl;

    return 0;
}